relay/relay.py

```diff
@@ -0,0 +1,364 @@
+    """Shut down the relay server
+
+    By default, Bottle runs the server until the user types Ctrl-C at the
+    terminal; however, neither raising KeyboardInterrupt nor using os.kill()
+    in a method callback actually results in an exit (the latter is mapped to
+    the former which then gets caught).  It might have something to do either
+    with the signal coming from the wrong thread or with it arriving _during_
+    the handling of a request.  (Calling exit() didn't work either -- that's
+    implemented as an exception, too, apparently.)  So, we need to send the
+    signal from outside the process, but using subproccess.run() doesn't work:
+    the signal interrupts the function's select(2) call as it waits for the
+    subprocess.  However, os.posix_spawn(), which creates an independent
+    process, seems to work!  (Note that the response is sent before the server
+    shuts down.)
+    """
+    logging.info("request to shut down")
+    os.posix_spawn("/usr/bin/kill", ("DIE.DIE.DIE", "-INT", str(os.getpid())), {})
+    return HTTPResponse(status=HTTPStatus.OK, body="Good bye!")
+
+
+@app.get("/<secret>/<file_id>")
+@validate_secret
+@click.pass_context
+def retrieve_file(context: click.Context, file_id: str) -> HTTPResponse:
+    """Send the requested file to the requester
+
+    Args:
+        context:  the Click context object
+                    - for access to the local files directory path
+        file_id:  the SHA256 hash of the file to be retrieved
+
+    Returns:
+        An HTTP response indicating the success of the download
+    """
+    logging.info('request to send file id "%s"', file_id)
+    return static_file(file_id, root=context.meta[CTX_DIRECTORY])
+
+
+@app.put("/<secret>/<file_id>")
+@validate_secret
+@click.pass_context
+def receive_file(context: click.Context, file_id: str) -> HTTPResponse:
+    """Receive the file sent by the requester
+
+    The file is uploaded and saved if it does not already exist locally, then
+    its SHA256 hash is calculated and compared to the file ID used in the
+    request.  If the match fails, the file is deleted and an error is returned.
+
+    Args:
+        context:  the Click context object
+                    - for access to the local files directory path
+        file_id:  the SHA256 hash of the file to be retrieved
+
+    Returns:
+        An HTTP response indicating the success of the upload
+    """
+    logging.info(
+        'request to upload file id "%s", disk %s',
+        file_id,
+        get_disk_utilization_str(context.meta[CTX_DIRECTORY]),
+    )
+
+    if not 0 < request.content_length <= FILE_MAX_SIZE:
+        return HTTPResponse(
+            status=HTTPStatus.BAD_REQUEST,
+            body=f"Content-Length ({request.content_length}) "
+            f"must be greater than zero and less than {FILE_MAX_SIZE}.",
+        )
+
+    remove_file = False
+    bytes_remaining = request.content_length
+    hash_sha256 = hashlib.sha256()
```

Comment:
Yeah, this has been nettling me, as well.  We should discuss this in real-time.

Reply:
I don't disagree with keeping the relay API and storage "minimal" but let's avoid deliberately making it "more minimal than necessary". At a minimum we need a name and resource ID on the server side. The resource ID can be generated as an MD5 on intake, but it's a waste of time to generate both. We can think about phasing out the old MD5, but that affects the Agent side as well, and I don't think either that or a generalization to handle multiple resource ID hashes should be critical for 1.0... while the relay definitely is.

As you commented elsewhere, we can support metadata on the server side `pull` API, and I intend to. However, there's a broad gap between origination of the data and pulling it to the server. Basically, the way the relay is set up, the artifacts you hold are identified only by a SHA256-based URI; and now on the dashboard they're supposed to figure out what metadata they meant to have for each SHA256.

The new `pbench-results-move --relay <uri>` (or however it looks) will report the set of URIs from which the data is available to the server, and those messages *could* associate each with the original Agent tarball names, but the `stdout` lines will be long and unwieldy, and we're basically requiring them to manually create a translation key they can go back to later... which doesn't sound like a very good UI to me.

As I mentioned before, a big challenge with `pbench-results-move` is that it may act on an arbitrary set of collected runs that happen to be present; right now it tells you when it's pushed each one, but doesn't really identify them, and there's no mechanism to associate distinct metadata with each. Which was "OK" for the old `--user` and maybe even for `--prefix` (though the later is squishier), but probably not very good for metadata. We could make `pbench-results-move` more "interactive", but that's less scriptable. Similarly we could pull out a command that would package and push individual runs by name or some other handle, but that requires a directory command to tell you what's there (and identify the handles), and it becomes a much more complicated UI. I don't have any good ideas to deal with this. *But* if we could, being able to specify the run metadata at "push" time would be better than deferring it.

Once we get that far, the best would be to add `--metadata` to the *benchmark scripts*, like `pbench-user-benchmark --config test --metadata "global.data:oc" -- sleep 10`, and propagate that in the run directory somewhere that `pbench-results-move` would read. But we still need a way to propagate that through to the server intake. (Yeah, maybe we could just put it in `metadata.log`, although that's a bit annoying as that's a permanent artifact of the dataset, and it's going to be redundant.)

Reply:
Let's take this up in Discussion #3414.